network/dag/publisher.go

```diff
@@ -179,24 +178,47 @@ func (s *replayingDAGPublisher) publishTransaction(ctx context.Context, transact
 		return false
 	}
 
-	// TODO: Now calls TransactionAddedEvent and TransactionPayloadAddedEvent after checken whether the payload is present.
-	//       This will need to be changed: TransactionAddedEvent must be called regardless whether the payload is present or not (e.g. top of this function).
-	//       However, when doing that at this moment, TransactionAddedEvent might be published multiple times for transactions which payload is not present the first time.
-	//       This is generally the case during operation when new transactions are received over the network.
-	//       Since not all subscribers are guaranteed to be idempotent at this time, they might break if we introduce it at this moment in time.
-	//       So after all subscribers are made idempotent, TransactionAddedEvent must be published regardless of the payload is present or not.
-	//       This is to accommodate syncing DAGs even when receiving a TX with a detached payload, or a private TX not meant for the local node.
-	for eventType := range s.subscribers {
-		for _, payloadType := range []string{transaction.PayloadType(), AnyPayloadType} {
-			receiver := s.subscribers[eventType][payloadType]
-			if receiver == nil {
-				continue
-			}
-			if err := receiver(transaction, payload); err != nil {
-				log.Logger().Errorf("Transaction subscriber returned an error (ref=%s,type=%s): %v", transaction.Ref(), transaction.PayloadType(), err)
-			}
+	s.emitEvent(TransactionPayloadAddedEvent, transaction, payload)
+
+	return true
+}
+
+func (s *replayingDAGPublisher) emitEvent(eventType EventType, transaction Transaction, payload []byte) {
+	for _, payloadType := range []string{transaction.PayloadType(), AnyPayloadType} {
+		subs := s.subscribers[eventType]
+		if subs == nil {
+			continue
+		}
+		receiver := subs[payloadType]
+		if receiver == nil {
+			continue
+		}
+		if err := receiver(transaction, payload); err != nil {
+			log.Logger().Errorf("Transaction subscriber returned an error (ref=%s,type=%s): %v", transaction.Ref(), transaction.PayloadType(), err)
 		}
 	}
+}
 
-	return true
+func (s *replayingDAGPublisher) replay() error {
+	log.Logger().Debug("Replaying DAG...")
+	s.publishMux.Lock()
+	defer s.publishMux.Unlock()
+
+	err := s.dag.Walk(context.Background(), func(ctx context.Context, tx Transaction) bool {
+		s.transactionAdded(ctx, tx)
+		payload, err := s.payloadStore.ReadPayload(ctx, tx.PayloadHash())
+		if err != nil {
+			log.Logger().Errorf("Error reading payload (tx=%s): %v", tx.Ref(), err)
+		}
+		if payload == nil {
+			return false
```

Comment:
what if you use the payload check to call payloadWritten? And always return true?

Reply:
> what if you use the payload check to call payloadWritten? And always return true?

maybe better to tackle that issue in the next PR.